common/main/java/com/couchbase/lite/internal/replicator/AbstractCBLWebSocket.java

```diff
@@ -649,17 +647,34 @@ private int getCodeForError(Throwable error) {
         return 0;
     }
 
-    @SuppressWarnings("PMD.AvoidRethrowingException")
-    private void setupBasicAuthenticator(@NonNull Map<?, ?> auth, @NonNull OkHttpClient.Builder builder) {
-        final Object username = auth.get(C4Replicator.REPLICATOR_AUTH_USER_NAME);
-        final Object password = auth.get(C4Replicator.REPLICATOR_AUTH_PASSWORD);
-        if (!((password instanceof String) && (username instanceof String))) { return; }
-
-        final String credentials = Credentials.basic((String) username, (String) password);
+    private void setupAuthentication(@NonNull OkHttpClient.Builder builder, @NonNull Map<?, ?> auth) {
+        String proxyCredentials = null;
+        final Object proxyUser = auth.get(C4Replicator.REPLICATOR_OPTION_PROXY_USER);
+        final Object proxyPass = auth.get(C4Replicator.REPLICATOR_OPTION_PROXY_PASS);
+        if (((proxyUser instanceof String) && (proxyPass instanceof String))) {
+            proxyCredentials = Credentials.basic((String) proxyUser, (String) proxyPass);
+        }
+        final String proxyCred = proxyCredentials;
+
+        String endpointCredentials = null;
+        if (C4Replicator.AUTH_TYPE_BASIC.equals(auth.get(C4Replicator.REPLICATOR_AUTH_TYPE))) {
+            final Object endptUser = auth.get(C4Replicator.REPLICATOR_AUTH_USER_NAME);
+            final Object endptPass = auth.get(C4Replicator.REPLICATOR_AUTH_PASSWORD);
+            if (((endptUser instanceof String) && (endptPass instanceof String))) {
+                endpointCredentials = Credentials.basic((String) endptUser, (String) endptPass);
+                forcePreAuth(builder, endpointCredentials);
+            }
+        }
+        final String endptCred = endpointCredentials;
 
-        builder.authenticator((route, resp) -> authenticate(resp, credentials));
+        if ((proxyCredentials != null) || (endpointCredentials != null)) {
+            builder.authenticator((route, resp) -> authenticate(resp, endptCred, proxyCred));
```

Comment:
Following this code sample, from the OkHttp 3.14 (the version we use) docs:
https://square.github.io/okhttp/3.x/okhttp/okhttp3/Authenticator.html

As I understand it, I will get this callback, preemptively, so that OkHttp has the credentials *before* it tries to connect.
I could be mistaken.

Reply:
I see. I think it could be a good idea if we do a manual test at least once to confirm that all are correct including the OkHttp-Preemptive challenge name.